armi/reactor/converters/axialExpansionChanger.py

```diff
@@ -182,6 +199,145 @@ def _isTopDummyBlockPresent(self):
+            else:
+                blockHeight = b.p.height
+            # set bottom of block equal to top of block below it
+            # if ib == 0, leave block bottom = 0.0
+            if ib > 0:
+                b.p.zbottom = self.linked.linkedBlocks[b][0].p.ztop
+            isDummyBlock = ib == (numOfBlocks - 1)
+            if not isDummyBlock:
+                ## grow all the components
+                for c in _getSolidComponents(b):
+                    growFrac = self.expansionData.getExpansionFactor(c)
+                    runLog.debug(
+                        msg="      Component {0}, growFrac = {1:.4e}".format(
+                            c, growFrac
+                        )
+                    )
+                    if thermal and c in self.expansionData.componentReferenceHeight:
+                        blockHeight = self.expansionData.componentReferenceHeight[c]
+                    if growFrac >= 0.0:
+                        c.height = (1.0 + growFrac) * blockHeight
+                    else:
+                        c.height = (1.0 / (1.0 - growFrac)) * blockHeight
+
+                    ## align upward expanding components
+                    # - except for control pins within control bundle (those are specially handled)
+                    if ib == 0:
+                        c.zbottom = 0.0
+                    else:
+                        if c.axiallyExpandsDownwards:
+                            componentsToExpandDown[b].append(c)
+                            # topAnchorForDownwardExpansion will overwrite itself
+                            # until at top most block with components that can expand downwards
+                            topAnchorForDownwardExpansion = b.p.ztop
+                            continue
+                        if c.hasFlags(Flags.CONTROL):
+                            continue
+                        if self.linked.linkedComponents[c][0] is not None:
+                            # use linked components below
+                            c.zbottom = self.linked.linkedComponents[c][0].ztop
+                        else:
+                            # otherwise there aren't any linked components
+                            # so just set the bottom of the component to
+                            # the top of the target component in the block below it
+                            targetCompForBlockBelow = (
+                                self.expansionData.getTargetComponent(
+                                    self.linked.linkedBlocks[b][0]
+                                )
+                            )
+                            c.zbottom = targetCompForBlockBelow.ztop
+                    c.ztop = c.zbottom + c.height
+
+        # align downward expanding components
+        # new c.height is set, but c.zbottom and c.ztop need to be set
+        blockList = list(componentsToExpandDown.keys())
+        componentPerBlockList = list(componentsToExpandDown.values())
```

Comment:
Any particular reason to convert this to a `list` instead of using `componentsToExpandDown` directly?

Reply:
So this is just what I came up with at the time. If you can think of something prettier, I'm all for it! Here are the needs:
- `componentToExpandDown` is a dict where the keys are blocks and values are a list of components that need to be expanded downwards on that block. The dictionary contains (key, value) pairs from bottom to top in the assembly. 
- This loop, which is step 2 in the attached PDF, has to run from top to bottom. 
- Since dict.keys()/values() returns a view object, I chose to cast the keys and values to lists so they could iterated on in the fashion shown. 

If there is a better way to do this (there very likely is!) I'm all for it.

Reply:
This isn't really important, but I think this would work:

```
for ib, (b, compList) in reversed(list(enumerate(list(componentsToExpandDown.items())))):
    for c in compList:
        if ib == len(componentsToExpandDown) - 1:
            etc ...
```

Now that I've written it out it looks _less_ elegant than the original, lol. But, that's what I was thinking.

Reply:
The main thing I thought was that you could use `componentsToExpandDown[b]` instead of `componentPerBlockList[ib]`, but again, that doesn't really matter much.

Reply:
If there is consensus that it (or something like it) can improve the readability/maintainability, I'm 100% down to make any changes. 

@john-science any ideas here?

Reply:
No, I think it's fine the way it is. Using the dictionary looked better in my head than in reality.